// @echarts-readymade/core/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "COLOR_LIST", {
  enumerable: true,
  get: function get() {
    return _misc.COLOR_LIST;
  }
});
Object.defineProperty(exports, "ChartContext", {
  enumerable: true,
  get: function get() {
    return _ChartProvider.ChartContext;
  }
});
Object.defineProperty(exports, "ChartProvider", {
  enumerable: true,
  get: function get() {
    return _ChartProvider.ChartProvider;
  }
});
Object.defineProperty(exports, "buildChartOption", {
  enumerable: true,
  get: function get() {
    return _chartOptionBuilder.buildChartOption;
  }
});
exports["default"] = void 0;
Object.defineProperty(exports, "mergeOption", {
  enumerable: true,
  get: function get() {
    return _misc.mergeOption;
  }
});
Object.defineProperty(exports, "numberWithCommas", {
  enumerable: true,
  get: function get() {
    return _misc.numberWithCommas;
  }
});
Object.defineProperty(exports, "truncate", {
  enumerable: true,
  get: function get() {
    return _misc.truncate;
  }
});

var _ChartProvider = require("./ChartProvider");

var _misc = require("./misc");

var _chartOptionBuilder = require("./chartOptionBuilder");

var _default = {
  ChartProvider: _ChartProvider.ChartProvider,
  ChartContext: _ChartProvider.ChartContext,
  mergeOption: _misc.mergeOption,
  buildChartOption: _chartOptionBuilder.buildChartOption,
  numberWithCommas: _misc.numberWithCommas,
  COLOR_LIST: _misc.COLOR_LIST,
  truncate: _misc.truncate
};
exports["default"] = _default;
